### DISCUSSION TITLE
Use shortcode to locate a named shape inside an html-field?

### DISCUSSION THREAD

**Discussion by bashuss** [OrchardCMS/OrchardCore#11578] [Q&A] [answered]
**My situation**
I would like to create a ContentType that has an image collection, that can either be rendered as carousel before the html content as a standard or optional at a user defined place within an Html field.

**Idea**
I thought about placing the shape of the carousel inside a Property of a Model (within the template for the ContentType), which could then be access by a shortcode to be rendered at the location of the shortcode.

This just does not seem to work, as I either do not pass the shape correctly or not to a location, where it can be accessed by the shortcode. Or my definition of the shortcode is not correct.

The ideal usage inside the HTML could be like this: 
_[shape_locator name="Carousel"]_
or even
_[carousel]_

with shape_locator being defined like this:
_{{ Model.Properties[Args.name] | shape_render }}_
or
_{{ Model.Properties["Carousel"] | shape_render }}_

Inside the template of my ContentType I would pass the shape like this:
_{{ assign myCarousel = "CarouselShape" | shape_new Media: Model.Content.MyType-Carousel }}
{% shape_add_properties Model.Content.MyType-MyHtmlField Carousel: myCarousel %}_

I tried some other ideas to pass the data, but non of the with success.
Any idea how to achieve this?

I might get it working by already rendering the shape and pass the rendered text, but the call to the shortcode would be less elegant (bad for use by external authors).

**Comment by ns8482e** [comment]
`{{ Args.name | shape_render }}` should work

**Reply by bashuss** [reply]
It does not work here - if I understand correctly what you mean.

I have still Version 1.1. of Orchard Core - which may be the reason. I hope I'll have the time soon for .NET Framework Migration.

But can you please elaborate? 
Did you mean it this way?

The ShortCode [shape_locator] is defined like this:
_{{ Args.name | shape_render }}_

Inside the template of my ContentType I pass the shape like this:
_{{ assign myCarousel = "CarouselShape" | shape_new Media: Model.Content.MyType-Carousel }}
{% shape_add_properties Model.Content.MyType-MyHtmlField Carousel: myCarousel %}_

Inside the HTML of MyHtmlField I call the ShortCode like this:
_[shape_locator name="Carousel"]_

**Comment by jtkech** [comment, marked as answer]
At least for `assign` you shoukd use `{% assign ... %}`.

**Reply by bashuss** [reply]
yes, right ... but it was just a mistake in the post here - it is correct in my code

**Reply by bashuss** [reply]
Exact copy from the ContentType template - out-commented part just to check, if render of carousel works and it does:

{% assign HTML=Model.Content.BlogPost-InhaltHtml %} 

{{ Model.Content.ContentsMetadata | shape_render }}

{% if Model.ContentItem.Content.BlogPost.Carousel.Paths.size > 0 %}
  {% assign carousel = "CarouselShape" | shape_new: Media: Model.Content.BlogPost-Carousel %}
  {% shape_add_properties HTML Carousel: carousel %}
  
  {% comment %}
  {{ carousel | shape_render }}
  {% endcomment %}

{% endif %}

{{ HTML | shape_render }}
{{ Model.Content.BlogPost-Category | shape_render }}
{{ Model.Content.BlogPost-Tags | shape_render }}

**Reply by ns8482e** [reply]
> {% assign carousel = "CarouselShape" | shape_new: Media: Model.Content.BlogPost-Carousel %}

This line is `IShape` object not string. However your short code  `[shape_locator name="Carousel"]`  send string argument.

**Reply by bashuss** [reply]
Yes - if you look at my first Code in the original message, the idea is, that the author of the blog does not have to know the internal workings. So I wanted to store the shape at a place from where it could be taken just by it's name. For this reason I used "shape_add_properties".
From your answer I assumed, that the inner workings of the shortcode are different.

I also tried passing the shape, which did not work, as it either was not passed into the HTML or it was not used correctly inside the shortcode - basically my Args inside the shortcode template ended up empty

**Reply by ns8482e** [reply]
Short code are like functions only parameters are accessible inside and local variables of  caller are not available to short codes

Here `assign carousel` is local to content  template. if it’s set as context object then only it would work.

Btw,  you always have access to content item and it’s properties like media collection fields, you can used them to build shape and render within short code

**Reply by bashuss** [reply]
Thanks! I just saw the Context object that works inside the shortcodes. What it does was not completely clear and obvious from the documentation, but with this it works. I made it work with this and it will be a good solution for my case, though the shortcode will only work inside **BlogPost** Items like this.
Till now, there are no other uses, so this is perfectly acceptable solution.

My shortcode template, which can simply be used without parameters now (as simple as it goes - great):

_{% assign carousel = "CarouselShape" | shape_new: Media: Context.ContentItem.Content.**BlogPost**.Carousel %}
{{ carousel | shape_render }}_

I could check with if-clauses for different contenttypes and handle the differently - this should be enough for now.
Thanks to all of you!
